File: PathoFroh/src/main/java/com/patho/main/service/StainingPrototypeService.java
package com.patho.main.service;

import com.patho.main.model.StainingPrototype;
import com.patho.main.model.StainingPrototypeDetails;
import com.patho.main.repository.jpa.StainingPrototypeDetailsRepository;
import com.patho.main.repository.jpa.StainingPrototypeRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Propagation;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;
import java.util.Optional;

@Service
@Transactional
public class StainingPrototypeService extends AbstractService {

    @Autowired
    private StainingPrototypeRepository stainingPrototypeRepository;

    @Autowired
    private StainingPrototypeDetailsRepository stainingPrototypeDetailsRepository;

    public StainingPrototype addOrUpdate(StainingPrototype p, List<StainingPrototypeDetails> removeDetails) {

        p = stainingPrototypeRepository.save(p,
                resourceBundle.get(p.getId() == 0 ? "log.settings.staining.new" : "log.slide.edit.textUpdate", p));

        for (StainingPrototypeDetails stainingPrototypeDetails : removeDetails) {
            stainingPrototypeDetailsRepository.delete(stainingPrototypeDetails);
        }

        return p;
    }

    public StainingPrototype incrementPriorityCounter(StainingPrototype prototype) {
        Optional<StainingPrototype> p = stainingPrototypeRepository.findById(prototype.getId());

        if (p.isPresent()) {
            p.get().setPriorityCount(p.get().getPriorityCount() + 1);
            return stainingPrototypeRepository.save(p.get(), resourceBundle
                    .get("log.settings.staining.priority.increment", p.get().getName(), p.get().getPriorityCount()));
        }

        return prototype;
    }

    /**
     * Tries to delete the stainingprototype, if not possible the prototype will be
     * deleted
     *
     * @param p
     */
    @Transactional(propagation = Propagation.NEVER)
    public boolean deleteOrArchive(StainingPrototype p) {
        try {
            stainingPrototypeRepository.delete(p, resourceBundle.get("log.settings.staining.deleted", p.getName()));
            return true;
        } catch (Exception e) {
            archive(p, true);
            return false;
        }
    }

    /**
     * Archived or dearchives a prototype
     *
     * @param p
     * @param archive
     * @return
     */
    public StainingPrototype archive(StainingPrototype p, boolean archive) {
        p.setArchived(archive);
        return stainingPrototypeRepository.save(p, resourceBundle
                .get(archive ? "log.settings.staining.archived" : "log.settings.staining.dearchived", p.getName()));
    }
}

File: src/main/kotlin/com/patho/main/service/StainingPrototypeService.kt
package com.patho.main.service

import com.patho.main.model.interfaces.ListOrder
import com.patho.main.model.preset.ListItem
import com.patho.main.model.preset.ListItemType
import com.patho.main.model.preset.StainingPrototype
import com.patho.main.repository.jpa.StainingPrototypeRepository
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Propagation
import org.springframework.transaction.annotation.Transactional
import java.util.*

@Service()
open class StainingPrototypeService @Autowired constructor(
        private val stainingPrototypeRepository: StainingPrototypeRepository) : AbstractService() {

    /**
     * Adds or updates a new staining prototype
     */
    @Transactional
    open fun addOrUpdate(prototype: StainingPrototype): StainingPrototype {
        return stainingPrototypeRepository.save(prototype,
                resourceBundle[if (prototype.id == 0L) "log.settings.staining.new" else "log.slide.edit.textUpdate", prototype])
    }

    /**
     * Increases the priority counter of a preset
     */
    @Transactional
    open fun incrementPriorityCounter(prototype: StainingPrototype): StainingPrototype {
        val p: Optional<StainingPrototype> = stainingPrototypeRepository.findById(prototype.id)
        if (p.isPresent) {
            p.get().priorityCount = p.get().priorityCount + 1
            return stainingPrototypeRepository.save(p.get(), resourceBundle["log.settings.staining.priority.increment", p.get().name, p.get().priorityCount])
        }
        return prototype
    }

    /**
     * Tries to delete the staining prototype, if not possible the prototype will be
     * deleted
     */
    @Transactional(propagation = Propagation.NEVER)
    open fun deleteOrArchive(prototype: StainingPrototype): Boolean {
        return try {
            stainingPrototypeRepository.delete(prototype, resourceBundle.get("log.settings.staining.deleted", prototype.name))
            true
        } catch (e: Exception) {
            archive(prototype, true)
            false
        }
    }

    /**
     * Archived or dearchives a prototype
     */
    @Transactional
    open fun archive(p: StainingPrototype, archive: Boolean): StainingPrototype? {
        p.archived = archive
        return stainingPrototypeRepository.save(p, resourceBundle
                .get(if (archive) "log.settings.staining.archived" else "log.settings.staining.dearchived", p.name))
    }

    /**
     * Reorganizes and persists the list
     */
    @Transactional
    open fun updateReorderedList(stainingPrototypes: List<StainingPrototype>): List<StainingPrototype> {
        ListOrder.reOrderList(stainingPrototypes)
        return stainingPrototypeRepository.saveAll(stainingPrototypes, resourceBundle["log.settings.material.list.reoder"])
    }
}